source/geod24/concurrency.d

```diff
@@ -925,35 +925,45 @@ protected:
     static class InfoFiber : Fiber
     {
         ThreadInfo info;
+        FiberCondition cond;
 
         this(void delegate() op, size_t sz = 512 * 1024) nothrow
         {
+            cond = null;
             super(op, sz);
         }
     }
 
     protected class FiberCondition
     {
+        MonoTime limit = MonoTime.init;
+
+        private void registerToInfoFiber() nothrow
+        {
+            /*
+            *  Register the FiberCondition to InfoFiber so that FiberScheduler can check if condition is met
+            */
+            auto infoF = cast(InfoFiber) Fiber.getThis();
+            assert(infoF !is null, "This Fiber does not belong to FiberScheduler");
+            assert(infoF.cond is null, "This Fiber already has a registered FiberCondition");
+            infoF.cond = this;
+        }
+
         void wait() nothrow
```

Comment:
Could we pass the `InfoFiber` as an argument to avoid relying on global state in `registerToInfoFiber`.
Not to mentioned, we would be able to inline it inside `wait`.

Reply:
Who would be supplying the `InfoFiber` ? 

Relying on `Fiber.getThis()` helps with a certain scenario. `registerToInfoFiber` binds the current Fiber and itself at the time of `wait` call. This enables multiple Fibers to wait on same semaphore, if we ever need that. I made it into a function to reduce duplication since we have overloaded `wait`. We can also just drop `wait()` and have a single `wait (Duration period = 0)` to handle the infinite wait case as well.

Reply:
> Who would be supplying the `InfoFiber` ?

As a first step, you could just make it a default argument.

> This enables multiple Fibers to wait on same condition, if we ever need that.

So, here's my 2cts on it: Fibers are thread-local and *never* moved across threads. Each thread would have one or more scheduler (e.g. a task running in a scheduler might have its own scheduler as a task pool). A thread should *only* block if all its fibers are blocked. Otherwise it can schedule non-blocked fiber.
So there is a possibility that fibers block on the same condition, but that might be handled at a higher level, and fibers are likely to have their own condition.

> We can also just drop wait() and have a single `wait (Duration period = 0)` to handle the infinite wait case as well.

That'd be sweet. Maybe another PR though.

Reply:
> Each thread would have one or more scheduler (e.g. a task running in a scheduler might have its own scheduler as a task pool).

That should be possible with the current implementation.

> A thread should only block if all its fibers are blocked. Otherwise it can schedule non-blocked fiber.

That would be the ideal case, yes. But, if we want that level of control, I suppose we would need a fiber version of mutex, condvar etc, a more complex scheduler and run everything strictly in fibers.